sdk/core/core-http/lib/util/atomXmlHelper.ts

```diff
@@ -0,0 +1,338 @@
+      atomResponseInJson = await deserializeAtomXmlToJson(response.bodyAsText);
+      response.parsedBody = atomResponseInJson;
+    }
+  } catch (e) {
+    response.errorBody = {
+      error: { code: "ResponseNotInAtomXMLFormat" }
+    };
+  }
+
+  if (response.status < 200 || response.status > 300) {
+    if (response.errorBody == undefined) {
+      const HttpResponseCodes = Constants.HttpResponseCodes;
+      const statusCode = response.status;
+      if (!isKnownResponseCode(statusCode)) {
+        response.errorBody = {
+          error: { code: `UnrecognizedHttpResponseStatus: ${statusCode}` }
+        };
+      } else {
+        response.errorBody = { error: { code: HttpResponseCodes[statusCode] } };
+      }
+    }
+  }
+
+  // Transform the errorBody to a normalized one
+  const normalizedError = normalizeError(response.errorBody, response);
+  response.errorBody = normalizedError;
+
+  // Construct response with 'result' to be backward compatibile
+  const responseInCustomJson: any = {
+    error: response.errorBody,
+    response: response.parsedBody,
+    result: shouldParseResponse ? [] : undefined
+  };
+
+  if (responseInCustomJson.error == undefined) {
+    const result = shouldParseResponse ? parseAtomResult(atomResponseInJson) : undefined;
+    if (result) {
+      if (Array.isArray(result)) {
+        result.forEach((entry: XMLResponseInJSON) => {
+          setName(entry, nameProperties);
+        });
+      } else {
+        setName(result, nameProperties);
+      }
+    }
+    responseInCustomJson.result = result;
+  }
+
+  response.parsedBody = responseInCustomJson;
+  return response;
+}
+
+function isKnownResponseCode(
+  statusCode: number
+): statusCode is keyof typeof Constants.HttpResponseCodes {
+  return !!(Constants.HttpResponseCodes as { [statusCode: number]: string })[statusCode];
+}
+
+/**
+ * @ignore
+ * Extracts the applicable entity name(s) from the URL based on the known structure
+ * and instantiates the corresponding name properties to the deserialized response
+ *
+ * For instance, following is the URL structure for when creating a rule
+ * `<namespace-component>/<topic-name>/Subscriptions/<subscription-name>/Rules/<rule-name>`
+ *
+ * @param entry
+ * @param nameProperties
+ */
+function setName(entry: XMLResponseInJSON, nameProperties: any): any {
+  if (entry[Constants.ATOM_METADATA_MARKER]) {
+    const parsedUrl = new URL(entry[Constants.ATOM_METADATA_MARKER].id);
+
+    const parts = parsedUrl.pathname!.split("/");
+
+    for (let i = 0; i * 2 < parts.length - 1; i++) {
+      entry[nameProperties[i]] = parts[i * 2 + 1];
+    }
+  }
+}
+
+/**
+ * @ignore
+ * Utility to help construct the normalized error object based on given `errorBody`
+ * data and other data present in the received `response` object.
+ *
+ * @param errorBody
+ * @param response
+ */
+function normalizeError(errorBody: any, response: HttpOperationResponse): any {
```

Comment:
I am slightly confused by the purpose of this method.
From what I can see in this PR, `errorBody` is only set in the `deserializeAtomXmlResponse` method where it is always set in the below format
```
{
  error: {
        code: errorCode
    }
}
```

The errorBody would be either undefined or in the above format.
If so, then what are we trying to normalize here?

Reply:
That's only in case when the errorBody is `undefined`. When it's not, the error gets normalized based on the content returned by the service.

Reply:
I couldn't find the code that updates the `errorBody` with some kind of content returned by the service. Can you please point me to it? 

I could only find the 3 places in `deserializeAtomXmlResponse` where errorBody gets set by us using the format in my comment above
- ResponseNotInAtomXMLFormat
- UnrecognizedHttpResponseStatus
- Known Status Code

Reply:
The error information gets returned by the service in the response itself, looks like this got lost during refactoring.

Addressed this and other comments with code changes - mostly with tests added as part of #4922, this resulted in cascading of some more refactoring.
Main points for context
- the error information returned on response for get API is inconsistent for queues/topics vs subscription and rules.
- `shouldParseResponse` flag is not needed if ignoring differences between `null` vs `undefined` for the `result`

Reply:
Resolving as this code has been moved out of core-http and is covered in #4922